src/main/java/org/openstreetmap/atlas/checks/validation/tag/RoadNameGapCheck.java

```diff
@@ -0,0 +1,201 @@
+     */
+    private static final Predicate<AtlasObject> VALID_HIGHWAY_TAG = object -> Validators
+            .hasValuesFor(object, HighwayTag.class)
+            && Validators.isOfType(object, HighwayTag.class, HighwayTag.TERTIARY,
+                    HighwayTag.PRIMARY, HighwayTag.SECONDARY, HighwayTag.MOTORWAY,
+                    HighwayTag.TRUNK);
+
+    /**
+     * The default constructor that must be supplied. The Atlas Checks framework will generate the
+     * checks with this constructor, supplying a configuration that can be used to adjust any
+     * parameters that the check uses during operation.
+     *
+     * @param configuration
+     *            the JSON configuration for this check
+     */
+    public RoadNameGapCheck(final Configuration configuration)
+    {
+        super(configuration);
+    }
+
+    /**
+     * This function will validate if the supplied atlas object is valid for the check.
+     *
+     * @param object
+     *            the atlas object supplied by the Atlas-Checks framework for evaluation
+     * @return {@code true} if this object should be checked
+     */
+    @Override
+    public boolean validCheckForObject(final AtlasObject object)
+    {
+        return object instanceof Edge && Edge.isMasterEdgeIdentifier(object.getIdentifier())
+                && !HighwayTag.isLinkHighway(object) && VALID_HIGHWAY_TAG.test(object)
+                && !JunctionTag.isRoundabout(object);
+    }
+
+    /**
+     * This is the actual function that will check to see whether the object needs to be flagged.
+     *
+     * @param object
+     *            the atlas object supplied by the Atlas-Checks framework for evaluation
+     * @return an optional {@link CheckFlag} object that
+     */
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        final Edge edge = (Edge) object;
+        final Set<Edge> inEdges = edge.inEdges().stream().filter(this::validCheckForObject)
+                .collect(Collectors.toSet());
+        final Set<Edge> outEdges = edge.outEdges().stream().filter(this::validCheckForObject)
+                .collect(Collectors.toSet());
+
+        if (inEdges.isEmpty() || outEdges.isEmpty())
+        {
+            return Optional.empty();
+        }
+
+        final Set<String> matchingInAndOutEdgeNames = getMatchingInAndOutEdgeNames(inEdges,
+                outEdges);
+
+        if (matchingInAndOutEdgeNames.isEmpty())
+        {
+            // There is no pair of inedge and out edge with same name.
+            return Optional.empty();
+        }
+
+        // Create flag when we have in edge and out edge with same name but intermediate edge
+        // doesn't have a name.
+        if (!edge.getName().isPresent())
+        {
+            return Optional.of(
+                    createFlag(object, this.getLocalizedInstruction(0, edge.getOsmIdentifier())));
+        }
+        final Optional<String> edgeName = edge.getName();
+
```

Comment:
The code in this if condition is handling the 2 cases. Please take a look at getMatchingInAndOutEdgeNames() method and also finaMatchingEdgeNameWithAnyInEdgeOrOutEdge() method.